app/src/main/java/org/mozilla/focus/biometrics/BiometricAuthenticationHandler.kt

```diff
@@ -0,0 +1,181 @@
+        } catch (err: Exception) {
+            when (err) {
+                is NoSuchAlgorithmException,
+                is NoSuchProviderException ->
+                    throw RuntimeException("Failed to get an instance of KeyGenerator", err)
+                else -> throw err
+            }
+        }
+
+        createKey(DEFAULT_KEY_NAME, false)
+
+        val defaultCipher: Cipher
+        try {
+            defaultCipher = Cipher.getInstance(KeyProperties.KEY_ALGORITHM_AES + "/"
+                    + KeyProperties.BLOCK_MODE_CBC + "/"
+                    + KeyProperties.ENCRYPTION_PADDING_PKCS7)
+        } catch (err: Exception) {
+            when (err) {
+                is NoSuchAlgorithmException,
+                is NoSuchPaddingException ->
+                    throw RuntimeException("Failed to get an instance of Cipher", err)
+                else -> throw err
+            }
+        }
+
+        if (initCipher(defaultCipher, DEFAULT_KEY_NAME)) {
+            cryptoObject = FingerprintManagerCompat.CryptoObject(defaultCipher)
+        }
+
+        ProcessLifecycleOwner.get().lifecycle.addObserver(this)
+    }
+
+    private fun initCipher(cipher: Cipher, keyName: String): Boolean {
+        try {
+            keyStore?.load(null)
+            val key = keyStore?.getKey(keyName, null) as SecretKey
+            cipher.init(Cipher.ENCRYPT_MODE, key)
+            return true
+        }  catch (err: Exception) {
+            when (err) {
+                is KeyPermanentlyInvalidatedException -> return false
+                is KeyStoreException,
+                is CertificateException,
+                is UnrecoverableKeyException,
+                is IOException,
+                is NoSuchAlgorithmException,
+                is InvalidKeyException -> throw RuntimeException("Failed to init cipher", err)
+                else -> throw err
+            }
+        }
+    }
+
+    private fun createKey(keyName: String, invalidatedByBiometricEnrollment: Boolean?) {
+        try {
+            keyStore?.load(null)
+
+            val builder = KeyGenParameterSpec.Builder(keyName,
+                    KeyProperties.PURPOSE_ENCRYPT).setBlockModes(KeyProperties.BLOCK_MODE_CBC)
+                    .setUserAuthenticationRequired(true)
+                    .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_PKCS7)
+
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+                builder.setInvalidatedByBiometricEnrollment(invalidatedByBiometricEnrollment!!)
+            }
+
+            keyGenerator?.init(builder.build())
+            keyGenerator?.generateKey()
+
+        } catch (err: Exception) {
+            when (err) {
+                is NoSuchAlgorithmException,
+                is InvalidAlgorithmParameterException,
+                is CertificateException,
+                is IOException -> throw RuntimeException(err)
+                else -> throw err
```

Comment:
\[detekt\] ERROR: \[exceptions/RethrowCaughtException\] Do not rethrow a caught exception of the same type